chain/network/src/peer/peer_actor.rs

```diff
@@ -125,45 +131,116 @@ pub enum IOError {
     Send { tid: usize, message_type: String, size: usize },
 }
 
+#[derive(Debug)]
+pub(crate) enum StreamConfig {
+    Inbound,
+    Outbound { peer_id: PeerId },
+}
+
+/// PeerActor config, whose constructor may return an error.
+/// We need it because of actix which doesn't allow us
+/// to make PeerActor::new return an error.
+pub(crate) struct Config {
+    peer_addr: SocketAddr,
+    stream: tokio::net::TcpStream,
+    stream_config: StreamConfig,
+    connecting_status: ConnectingStatus,
+    network_state: Arc<NetworkState>,
+    force_encoding: Option<Encoding>,
+}
+
+impl Config {
+    pub fn new(
+        stream: tokio::net::TcpStream,
+        stream_config: StreamConfig,
+        force_encoding: Option<Encoding>,
+        network_state: Arc<NetworkState>,
+    ) -> anyhow::Result<Self> {
+        Ok(Self {
+            peer_addr: stream.peer_addr().context("stream.peer_addr()")?,
+            connecting_status: match &stream_config {
+                StreamConfig::Inbound => ConnectingStatus::Inbound(
+                    network_state
+                        .inbound_handshake_permits
+                        .clone()
+                        .try_acquire_owned()
+                        .context("too many connections in Connecting state")?,
+                ),
+                StreamConfig::Outbound { peer_id } => ConnectingStatus::Outbound(
+                    network_state
+                        .tier2
+                        .start_outbound(peer_id.clone())
```

Comment:
It obtains the permits to start a connection. That may fail, and that's the primary reason that it exists in the first place (otherwise I'd put it inside the PeerActor::new).